crawler/src/main.rs

```diff
@@ -0,0 +1,299 @@
+            listener_ip: Some(IpAddr::V4(Ipv4Addr::LOCALHOST)),
+            desired_listening_port: Some(4132),
+            max_connections: MAXIMUM_NUMBER_OF_PEERS as u16,
+            ..Default::default()
+        };
+
+        let pea2pea_node = Pea2PeaNode::new(Some(config)).await.unwrap();
+        let client_state = Default::default();
+        let node = Self {
+            synth_node: SynthNode::new(pea2pea_node, client_state),
+            known_network: Arc::new(KnownNetwork::default()),
+        };
+
+        node.enable_disconnect();
+        node.enable_handshake();
+        node.enable_reading();
+        node.enable_writing();
+
+        node
+    }
+
+    /// Spawns a task dedicated to broadcasting Ping messages.
+    pub fn send_pings(&self) {
+        let node = self.clone();
+        task::spawn(async move {
+            let genesis = Testnet2::genesis_block();
+            let ping_msg = ClientMessage::Ping(
+                MESSAGE_VERSION,
+                MAXIMUM_FORK_DEPTH,
+                node.node_type(),
+                node.state(),
+                genesis.hash(),
+                Data::Object(genesis.header().clone()),
+            );
+
+            loop {
+                if node.node().num_connected() != 0 {
+                    info!(parent: node.node().span(), "sending out Pings");
+                    node.send_broadcast(ping_msg.clone());
+                }
+                tokio::time::sleep(Duration::from_secs(PING_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Spawns a task dedicated to peer maintenance.
+    pub fn update_peers(&self) {
+        let node = self.clone();
+        task::spawn(async move {
+            loop {
+                // Disconnect from peers we have just crawled.
+                for addr in node.known_network.addrs_to_disconnect() {
+                    node.node().disconnect(addr).await;
+                }
+
+                // Connect to peers we haven't crawled in a while.
+                for addr in node.known_network.addrs_to_connect() {
+                    if !node.node().is_connected(addr) {
+                        let _ = node.node().connect(addr).await;
+                    }
+                }
+
+                info!(parent: node.node().span(), "Crawling the netowrk for more peers; asking peers for their peers");
+                node.send_broadcast(ClientMessage::PeerRequest);
+                tokio::time::sleep(Duration::from_secs(PEER_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Starts the usual periodic activities of a crawler node.
+    pub fn run_periodic_tasks(&self) {
+        self.send_pings();
+        self.update_peers();
+    }
+}
+
+/// Inbound message processing logic for the test nodes.
+#[async_trait::async_trait]
+impl Reading for Crawler {
+    type Message = ClientMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
```

Comment:
in order to achieve the best performance (and to reduce the required size of the read buffer), we should probably identify the type of the client message by its id and discard the unwanted messages (like blocks) altogether, rather than fully deserializing them; this is likely already partially achieved due to the small buffer size :slightly_smiling_face: